docs/qos/qos_config_documentation.adoc
[[qos-overview]]
= QoS database and settings documentation
Lukasz Budryk, Florian Jung
v2.0.0, 2020-05-11
:toc: left
:toclevels: 3

[[summary]]
== Abstract
This document covers all necessary information on how to set up QoS (quality of service) tests for the *NetTest*. It also contains a detailed description of all database relevant fields needed for these tests.
The basic concept of the *NetTest QoS-Tests* is the following:

* Dynamic design: all tests can be set up without changing the code of client and/or server:
** Tests can be turned on/off or fully edited whenever they need to be.
** Support for different languages can be added anytime
** Because of the real-time and on-demand evaluation old tests are being affected by all changes too (great advantage for example if a new language has been implemented, or a new test evaluation method has been added)

The second and third section: <<qos-test-result>> and <<database-qos_measurement_objective>> will cover all available json keys, used for test objectives as well as for all test results.

The fourth section: <<testscript>> covers the dynamic content by using a simple script.

The fifth section: <<examples>> has some step by step examples.

[[qos-test-result]]
== QoS test results

[[qos-test-result-over]]
=== QoS result overview
There are 8 different tests so far:

* <<qos-test-result-json-website,WEBSITE>>
* <<qos-test-result-json-http,HTTP PROXY>>
* <<qos-test-result-json-nontransproxy,NON TRANSPARENT PROXY>>
* <<qos-test-result-json-dns,DNS>>
* <<qos-test-result-json-tcp,TCP>>
* <<qos-test-result-json-udp,UDP>>
* <<qos-test-result-json-voip,VOICE OVER IP>>
* <<qos-test-result-json-traceroute,TRACEROUTE>>

All test objectives and results are stored as JSON key-value pairs. This chapter will cover all available JSON keys and their purpose. They are stored in the `configuration` database. Basically all test result (and objective) JSON keys are composed of 3 components:

 TESTTYPE_RESULTTYPE_LABEL

where:

* `TESTTYPE` defines the type of the test (see list below) and can have the following values (they represent the list below in the same order):
** `website`
** `http`
** `nontransproxy`
** `dns`
** `tcp`
** `udp`
** `voip`
** `traceroute`

* `RESULTTYPE` can be:
** `objective`
** `result`

+
`objecitve` means that this was a test parameter (objective) from the control server, where `result` is used to hold all test result values.

* `LABEL` is only a label to make the results more human readable

[[qos-test-result-json]]
=== JSON specification
Following these naming conventions the specification has been set up. Here are all JSON keys grouped by the test type:

[[qos-test-result-json-all]]
==== all tests
these key/value pairs can be accessed from all tests

    * `start_time_ns` => starting time of the test in ns relative to the speedtest start
    * `duration_ns` => duration of the test in ns

[[qos-test-result-json-website]]
==== website (= Website download test)
    * `website_objective_url` => the target URL of this test
    * `website_objective_timeout` => test timeout
    * `website_result_tx_bytes` => bytes transferred during the test
    * `website_result_rx_bytes` => bytes received during the test
    * `website_result_duration` => time needed to download and render the website in nanoseconds
    * `website_result_status` => the status code (http header), this can have the value -1 if the website was unreachable
    * `website_result_info` => information about the test procedure; it is generates by the client itself, depending on how the test ended; possible values:
            ** `OK` => test worked as expected
            ** `ERROR` => some error occured (most of the cases: unreachable target)
            ** `TIMEOUT` => timeout exceeded (see parameter: website_objective_timeout)

[[qos-test-result-json-http]]
==== http (= Http proxy test)
    * `http_objective_range` (_optional_) => defines the string that will be passed to the range request header
    * `http_objective_url` => target of this test
    * `http_result_header` => response header
    * `http_result_length` => content length
    * `http_result_hash` => checksum of the content, this field also contains information about the final result of the test
        It may contain the following values:
            ** 'any hexadecimal value' - the md5 checksum of the content that was downloaded
            ** `TIMEOUT` - the download timeout has been reached
            ** `ERROR` - another error occured (host not available, connection timeout, etc.)
    * `http_result_status` => status code in the response header (or -1 if there was no response)

[[qos-test-result-json-nontransproxy]]
==== nontransproxy (= Non transparent proxy test)
    * `nontransproxy_objective_request` => request string for this test
    * `nontransproxy_objective_port` => test port
    * `nontransproxy_objective_timeout` => timeout
    * `nontransproxy_result` => enum that represents the result status of this test. Possible values are:
            ** `OK` - the test was successful (=test execution; regardless of the test result)
            ** `TIMEOUT` - the download timeout has been reached
            ** `ERROR` - another error occured (host not available, connection timeout, etc.)
    * `nontransproxy_result_response` => response (echo from test server)

[[qos-test-result-json-dns]]
==== dns (= DNS test)
    * `dns_objective_host` => target host of this test
    * `dns_objective_dns_record` => dns record to request
    * `dns_objective_resolver` => dns resolver to be used for this test
    * `dns_objective_timeout` => dns query timeout in ns
    * `dns_result_duration` => time needed to complete the test in ns
    * `dns_result_info` => enum that represents the result status of this test. Possible values are:
        ** `OK` - the test was successful (=test execution; regardless of the test result)
        ** `TIMEOUT` - the dns query timeout has been reached
        ** `ERROR` - another error occured
    * `dns_result_status` => the query status; the most common values are:
        ** `NOERROR` => request completed without any error
        ** `NXDOMAIN` => non existent domain
    * `dns_result_entries_found` => number of entries found
    * `dns_result_entries` => result of this test containing all dns entries that were found (IMPORTANT: the value of this object is always an array, even if there is only one entry)
            an entry is composed of the following:
        ** `dns_result_ttl` => the time to live of the dns entry
        ** `dns_result_address` => the address this dns entry points to
        ** `dns_result_priority` => priority, if exists (as in MX or SRV record)

[[qos-test-result-json-tcp]]
==== tcp (= TCP incoming/outgoing test)
    * `tcp_objective_timeout` => test timeout
    * `tcp_objective_in_port` => port number used for the incoming test
    * `tcp_result_in` => enum:
        ** `OK` - incoming test succeeded
        ** `FAILED` - incoming test failed.
    * `tcp_result_in_response` => server message received after a connection was established
    * `tcp_objective_out_port` => port number used for the outgoing test
    * `tcp_result_out` => enum:
        ** `OK` - outgoing test succeeded
        ** `FAILED` - outgoing test failed.
    * `tcp_result_out_response` => response that the client received after sending an message to the test server

[[qos-test-result-json-udp]]
==== udp (= UDP incoming/outgoing test)
    * `udp_objective_timeout` => test timeout
    * `udp_objective_delay` => delay between packets (in ns)
    * `udp_objective_out_port` => port number used for the outgoing test
    * `udp_objective_out_num_packets` => the number of packets to be sent by the client
    * `udp_result_out_num_packets` => the number of packets received by the test server
    * `udp_result_out_packet_loss_rate` => outgoing packet loss rate
    * `udp_result_out_response_num_packets` => responses to outgoing packets
    * `udp_objective_in_port` => port number used for the incoming test
    * `udp_objective_in_num_packets` => the number of packets to be sent by the test server
    * `udp_result_in_num_packets` => the number of packets received by the client
    * `udp_result_in_response_num_packets` => responses to incoming packets received from server
    * `udp_result_in_packet_loss_rate` => incoming packet loss rate

[[qos-test-result-json-voip]]
==== voip (= Voice over IP test)

[IMPORTANT]
The *VoIP* test uses the RTP protocol as defined in RFC 3550 <<citation-1,[1]>>.

    * `voip_objective_delay` => (_optional_) delay between packets in ns, default: 20000000ns (=20ms)
    * `voip_objective_timeout` => (_optional_) test timeout, default: 3000000000ns (=3000ms)
    * [[voip_objective_payload]]`voip_objective_payload` => (_optional_) payload type, as defined in RFC 3551 <<citation-2,[2]>>, supported payload types and their values can be found in the table below. The relevant field is *"Payload type"*. The default value is: `0` (=PCMU).
+
[IMPORTANT]
Codecs with payload type "_dyn_" have no static payload type assigned and are only used with a dynamic payload type <<citation-2,[2]>>. These codecs are not supported by the VoIP test.
+
[format="csv", options="header", cols="<,^s,<,<"]
|===
Codec name, Payload type, Clock rate, Codec type
"PCMU", 0, 8000, AUDIO
"GSM", 3, 8000, AUDIO
"G723", 4, 8000, AUDIO
"DVI4_8", 5, 8000, AUDIO
"DVI4_16", 6, 16000, AUDIO
"LPC", 7, 8000, AUDIO
"PCMA", 8, 8000, AUDIO
"G722", 9, 8000, AUDIO
"L16_1", 10, 44100, AUDIO
"L16_2", 11, 44100, AUDIO
"QCELP", 12, 8000, AUDIO
CN,13, 8000, AUDIO
MPA,14, 90000, AUDIO
G728,15, 8000, AUDIO
DVI4_11,16, 11025, AUDIO
DVI4_22,17, 22050, AUDIO
G729,18, 8000, AUDIO
"G726_40",_dyn_, 8000, AUDIO
"G726_32",_dyn_, 8000, AUDIO
"G726_24",_dyn_, 8000, AUDIO
"G726_16",_dyn_, 8000, AUDIO
G729D,_dyn_, 8000, AUDIO
G729E,_dyn_, 8000, AUDIO
"GSM_EFR",_dyn_, 8000, AUDIO
L8,_dyn_,_variable_, AUDIO
RED,_dyn_,_variable_, AUDIO
VDVI,_dyn_,_variable_, AUDIO
CELB,25, 90000, VIDEO
JPEG,26, 90000, VIDEO
NV,28, 90000, VIDEO
H261,31, 90000, VIDEO
MPV,32, 90000, VIDEO
MP2T,33, 90000, BOTH
H263,34, 90000, VIDEO
"H263_1998",_dyn_, 90000, VIDEO
|===

    * `voip_objective_in_port` => the port for the incoming voice stream
    * `voip_objective_out_port` => the port for the outgoing voice stream
    * `voip_objective_call_duration` => (_optional_) duration of the simulated call, default: 1000000000ns (=1000ms)
    * `voip_objective_bits_per_sample` => (_optional_) bits per sample, default: 8
    * `voip_objective_sample_rate` => (_optional_) the sample rate in _Hz_, default: 8000
    * `voip_result_status` => the test result, enum:
    ** `OK` - the test was successful (=test execution; regardless of the test result)
    ** `TIMEOUT` - the test timeout has beed reached
    ** `ERROR` - another error occured
    * incoming voice stream results (client side):
    ** `voip_result_in_short_seq` => the shortest correct packet sequence (fewest number of packets in correct order)
    ** `voip_result_in_long_seq` => the longest correct packet sequence (most number of packets in correct order)
    ** `voip_result_in_max_jitter` => the max jitter in ns
    ** `voip_result_in_mean_jitter` => the mean jitter in ns
    ** `voip_result_in_skew` => the skew in ns
    ** `voip_result_in_num_packets` => number of packets received
    ** `voip_result_in_max_delta` => highest delay between received packets
    ** `voip_result_in_sequence_error` => number of sequence errors (packets out of order)
    * outgoing voice stream results:
    ** `voip_result_out_short_seq` => the shortest correct packet sequence (fewest number of packets in correct order)
    ** `voip_result_out_long_seq` => the longest correct packet sequence (most number of packets in correct order)
    ** `voip_result_out_max_jitter` => the max jitter in ns
    ** `voip_result_out_mean_jitter` => the mean jitter in ns
    ** `voip_result_out_skew` => the skew in ns
    ** `voip_result_out_max_delta` => highest delay between received packets
    ** `voip_result_out_num_packets` => number of packets received
    ** `voip_result_out_sequence_error` => number of sequence errors (packets out of order)

[[qos-test-result-json-traceroute]]
==== traceroute (= Traceroute test)
    * `traceroute_objective_host` => the target host
    * `traceroute_objective_timeout` => test timeout
    * `traceroute_objective_max_hops` => (_optional_) max hops allowed, default: 30
    * `traceroute_result_details` => a detailed list of the route (IMPORTANT: the value of this object is always an array, even if there is only one entry). An entry is composed of:
        ** `host` => host ip (and name if available)
        ** `time` => the time needed to reach the host
    * `traceroute_result_status` => enum:
        ** `OK` - test succeeded
        ** `TIMEOUT` - timeout has been reached.
        ** `MAX_HOPS_EXCEEDED` - max hops (see `traceroute_objective_max_hops`) has been exceeded before the target could be reached.
        ** `FAILED` - test failed (some other error occured during the test)
    * `traceroute_result_hops` => hops needed to reach the target host.


[[database-qos_measurement_objective]]
== QoS test objectives
The purpose of this is to provide tests (in detail: test parameters and test objectives) for all clients.

Fields:

    * `qos_test_uid`, type: _integer_ => uid
    * `type`, type: _qostest_ => the test type
    * `params`, type: _json_ => (see 3.1.1) test objectives.
    * `results`, type: _json_ => (see 3.1.3) expected test results.
    * `concurrency_group`, type: _integer_ => (see 3.1.5) tests that belong to the same group are executed simultaneously. This is also the order of test execution.
    * `description`, type: _text_ => (see 3.1.6) references the key of a text entry in the qos_test_desc table. This is the longer and more technical test summary.
    * `summary`, type: _text_ => (see 3.1.7) references the key of a text entry in the qos_test_desc table. This is the short test summary.

[[database-qos_test_objective-param]]
=== Field `param`
Test parameters are basically json keys without `TESTTYPE` and `RESULTTYPE` (where `RESULTTYPE` must be: `objective`). This means that for example a website test json key (as described in the section: <<qos-test-result-json>>) with the name: `website_objective_url` would be transformed to `url`. After a test has been executed by the client it adds the `TESTTYPE` and `RESULTTYPE` prefix to the parameters and sends them back to the server. The following parameters (objectives) are supported (sorted by test type):

    * <<qos-test-result-json-website>>, example and available parameters:

        {
            "url":"http://alladin.at",
            "timeout":10000000000
        }

    ** `url` => the target URL of this test (see `website_objective_url`)
    ** `timeout` => test timeout (see `website_objective_timeout`)

    * <<qos-test-result-json-http>>, example and available parameters:

        {
            "range":"bytes=0-999",
            "target":"https://www.rtr.at",
            "conn_timeout":5000000000,
            "download_timeout":15000000000
        }

    ** `range` => (_optional_) defines the string that will be passed to the range request header (see `http_objective_range`)
    ** `url` => target url of this test (see `http_objective_url`)
    ** `conn_timeout` => connection timeout in ns (no equivalent json key = not in result table)
    ** `download_timeout` => download timeout in ns (no equivalent json key = not in result table)

    * <<qos-test-result-json-nontransproxy>>, example and available parameters:

        {
            "port":"%RANDOM 50000 55000%",
            "request":"GET / HTTR/7.9"
        }

    ** `port` => test port (see `nontransproxy_objective_port`)
    ** `request` => request string for this test (see `nontransproxy_objective_request`)
    ** `timeout` => test timeout (see `nontransproxy_objective_timeout`)

    * <<qos-test-result-json-dns>>, example and available parameters:

        {
            "host":"rtr.at",
            "record":"MX",
            "resolver":"8.8.8.8"
        }

    ** `host` => target host of this test (see `dns_objective_host`)
    ** `record` =>  dns record to request (see `dns_objective_record`)
    ** `resolver` => (_optional_) dns resolver to be used for this test (see `dns_objective_resolver`) - if not set the standard system resolver is used
    ** `timeout` => (_optional_) dns query timeout in ns (see `dns_objective_timeout`) - default: 5000000000ns (=5000ms)

    * <<qos-test-result-json-tcp>>, example and available parameters:

        {
            "timeout":3000000000,
            "out_port":"%RANDOM 20000 40000%"
        }

    ** `timeout` => test timeout (see `tcp_objective_timeout`)
    ** `out_port` => port number used for the outgoing test (see `tcp_objective_out_port`)
    ** `in_port` => port number used for the incoming test (see `tcp_objective_in_port`)

[TIP]
It is possible to create incoming-only or outgoing-only tests. This means: either `out_port` or `in_port` or both parameters have to be set.

    * <<qos-test-result-json-udp>>, example and available parameters:

        {
            "in_port":"%RANDOM 10000 50000%",
            "timeout":2500000000,
            "in_num_packets":"%RANDOM 8 12%",
            "delay":500000000
        }

    ** `delay` => (_optional_) delay between packets in ns (see `udp_objective_delay`) - default: 300000000ns (=300ms)
    ** `timeout` => test timeout (see `udp_objective_timeout`)
    ** `in_port` => port number used for the incoming test (see `udp_objective_in_port`)
    ** `in_num_packets` => the number of packets for the incoming test (see `udp_objective_in_num_packets`)
    ** `out_port` => port number used for the outgoing test (see `udp_objective_out_port`)
    ** `out_num_packets` => the number of packets for the outgoing test (see `udp_objective_out_num_packets`)

[TIP]
It is possible to create incoming-only or outgoing-only tests. This means: either [`out_port` and `out_num_packets`] or [`in_port` and `in_num_packets`] or both parameter groups have to be set.

    * <<qos-test-result-json-voip>>, example and available parameters:

        {
            "in_port": "5060",
            "out_port": "5060",
            "timeout": "6000000000",
            "call_duration": "2000000000"
        }

    ** `delay` => (_optional_) delay between packets in ns, default: 20000000ns (=20ms)
    ** `timeout` => (_optional_) test timeout, default: 3000000000ns (=3000ms)
    ** `payload` => (_optional_) payload type, default: 0 (=PCMU). For a detailed documentation see: <<voip_objective_payload, VoIP: voip_objective_payload>>
    ** `out_port` => the port for the outgoing voice stream
    ** `in_port` => the port for the incoming voice stream
    ** `call_duration` => (_optional_) duration of the simulated call, default: 1000000000ns (=1000ms)
    ** `bits_per_sample` => (_optional_) bits per sample, default: 8
    ** `sample_rate` => (_optional_) the sample rate in _Hz_, default: 8000

    * <<qos-test-result-json-traceroute>>, example and available parameters:

        {
            "host": "google.com",
            "timeout": "35000000000"
        }

    ** `host` => the target host
    ** `timeout` => test timeout
    ** `max_hops` => (_optional_) max hops allowed, default: 30


[[database-qos_test_objective-result]]
=== Field `result`
This field contains the expected test results (from now on: 'ETR') and the behaviour in case of failure and/or success.

The reason that an 'ETR' is an array, is the method of evaluation: an 'ETR' can have multiple conditions. To explain the functionality of ETRs imagine the following scenario:

[abstract]
A non transparent proxy test is run on a random port between 1 and 25000. The request it will send is an erroneous HTTP request (`GET / HTTR 7.9`). After the test has finished we compare the result of the test to the request. If both are equal, then no proxy has changed the request, ergo no proxy has been detected. Which means that the test was successful in the QoS point of view. A second check is done to determine the port number classification (just a check if the random port was not above 1024). This has nothing to do with quality of service but will demonstrate the possibilities of the *NetTest QoS-Tests*. Here are the objectives:

[source,json]
    {
        "port":"%RANDOM 1 25000%",
    	"request":"GET / HTTR/7.9"
    }


website/clients/berec/settings.js
module.exports = {
    features: {
        "home": true,
        "tc": true,
        "nettest": true,
        "statistics": true,
        "map": true,
        "history": true,
        "testresults": true,
        "help": true,
        "opendata": true,
        "about": false,
        "footer": true,
        "settings": true,
        "docu": false
    },
    landing_page: "home",
    user_agent: "abc",
    languages: [
        "en",
    ],
    servers: {
        map: "",
        statistic: "",
        control: "http://localhost:8080/api/v1/"
    },
    keys: {
        "google": ""
    },
    classificationColorStyle: "background",
    colors: {
        groups: {
            pink: ["#da7883", "#d16b74", "#a65056", "#8a3d43"],
            orange: ["#ffce99", "#ffab65", "#ff8118", "#e67416"],
            purple: ["#e097ff", "#bb8ed8", "#a34ed8", "#69328b"],
            turquoise: ["#c8ebef", "#90bcbf", "#739c9e", "#64878a"],
            green: ["#bfdb2d", "#a6bf27", "#8a9f21", "#71821b"]
        },
        gauge: {
            arc_background: "#EFEFEF",
            arc_inner: "#921F56",
            arc_outer: "#29348A",
            fontName: "arial",
            font: "#FFFFFF"
        }
    },
    nettest: {
        custom_tc: false,
        tag: null,
        tests: {
            qos: false,
            ndt: true,
            rmbt: true
        }
    },
    map: {
        showLegend: true,
        view: {
            default_map: "google",
            map_provider: ["google", "googleSatellite", "googleTerrain", "googleHybrid", "osm"],
            position: [48.209, 16.37],
            zoom_initial: 12,
            zoom_min: 1,
            zoom_max: 20,
            hybrid: {
                zoom_level: 15
            }
        },
        filter_defaults: {
            map_options: "all/download"
        },
        options: {
            statistical_method: [],
            map_options: [],
            period: [],
            technology: [],
            operator: [],
            layer_type: [
                {value: "hybrid", name: "MAP.SELECT.LAYER.HYBRID", unit: "", "default": true},
                {value: "heatmap", name: "MAP.SELECT.LAYER.HEATMAP", unit: ""},
                {value: "points", name: "MAP.SELECT.LAYER.POINTS", unit: ""}
            ]
        }
    },
    opendata: {
        external: [
            
        ],
        allow_full_download: false,
        report_start_date: "2017/03/01"
    },
    statistics: {
        graphs: {
            list_filters: {

            },
            list_devices: {
                enabled: false
            },
            list_providers: {
                enabled: false
            },
            os_part: {
                enabled: true
            },
            technology_part: {
                enabled: true
            },
            measurements_per_provider_speed: {
                enabled: false
            },
            measurements_per_provider_over_time_speed: {
                enabled: true,
                show_box: 6,
                select_method: "random",
                select_count: 6
            },
            measurements_per_provider_over_time_count: {
                enabled: true
            },
            measurements_per_technology_over_time: {
                enabled: true
            }
        }
    },
    result_list: {
        history_graph_enabled: true
    },
    result: {
        keys: [
            {
                header: "RESULT.DETAIL.SECTION.PROVIDER.HEADER",
                icon: "f",
                values: [
                    {
                        key: "cat_technology"
                    },
                    {
                        key: "ip_anonym"
                    },
                    {
                        key: "country_geoip"
                    },
                    {
                        key: "network_name"
                    },
                    {
                        key: "network_type"
                    }
                ]
            },
            {
                header: "RESULT.DETAIL.SECTION.UNCATECORIZED.HEADER",
                icon: "m",
                values: [
                    {
                        key: "asn"
                    },
                    {
                        key: "bytes_download",
                        formats: [
                            {
                                format: "number",
                                factor: 0.000001
                            },
                            {
                                format: "speed"
                            }
                        ],
                        unit: "MB"
                    },
                    {
                        key: "bytes_upload",
                        formats: [
                            {
                                format: "number",
                                factor: 0.000001
                            },
                            {
                                format: "speed"
                            }
                        ],
                        unit: "MB"
                    },
                    {
                        key: "client_version"
                    },
                    {
                        key: "community"
                    },
                    {
                        key: "connection"
                    },
                    {
                        key: "country_asn"
                    },
                    {
                        key: "country_location"
                    },
                    {
                        key: "district"
                    },
                    {
                        key: "distance"
                    },
                    {
                        key: "download_classification"
                    },
                    {
                        key: "download_kbit",
                        formats: [
                            {
                                format: "number",
                                factor: 0.001
                            },
                            {
                                format: "speed"
                            }
                        ],
                        unit: "Mbit/s"
                    },
                    {
                        key: "duration_download_ms",
                        formats: [
                            {
                                format: "ping"
                            }
                        ],
                        unit: "ms"
                    },
                    {
                        key: "duration_upload_ms",
                        formats: [
                            {
                                format: "ping"
                            }
                        ],
                        unit: "ms"
                    },
                    {
                        key: "lat",
                        formats: [
                            {
                                format: "fixed",
                                places: 4
                            }
                        ]
                    },
                    {
                        key: "long",
                        formats: [
                            {
                                format: "fixed",
                                places: 4
                            }
                        ]
                    },
                    {
                        key: "loc_accuracy"
                    },
                    {
                        key: "loc_src"
                    },

                    {
                        key: "lte_rsrp",
                        unit: "dBm"
                    },

                    {
                        key: "lte_rsrq",
                        unit: "dBm"
                    },
                    {
                        key: "model"
                    },
                    {
                        key: "model_native"
                    },
                    {
                        key: "network_country"
                    },
                    {
                        key: "network_mcc_mnc"
                    },
                    {
                        key: "num_threads"
                    },
                    {
                        key: "num_threads_requested"
                    },
                    {
                        key: "num_threads_ul"
                    },
                    {
                        key: "open_test_uuid"
                    },
                    {
                        key: "ping_classification"
                    },
                    {
                        key: "ping_ms",
                        formats: [
                            {
                                format: "ping"
                            }
                        ],
                        unit: "ms"
                    },
                    {
                        key: "ping_variance",
                        formats: [
                            {
                                format: "ping"
                            }
                        ],
                        unit: "ms"
                    },
                    {
                        key: "platform"
                    },
                    {
                        key: "public_ip_as_name"
                    },
                    {
                        key: "product"
                    },
                    {
                        key: "province"
                    },
                    {
                        key: "provider_name"
                    },
                    {
                        key: "roaming_type"
                    },
                    {
                        key: "server_name"
                    },
                    {
                        key: "signal_classification"
                    },
                    {
                        key: "signal_strength",
                        unit: "dBm"
                    },
                    {
                        key: "sim_country"
                    },
                    {
                        key: "test_duration",
                        unit: "s"
                    },
                    {
                        key: "test_if_bytes_download",
                        formats: [
                            {
                                format: "number",
                                factor: 0.000001
                            },
                            {
                                format: "speed"
                            }
                        ],
                        unit: "MB"
                    },
                    {
                        key: "test_if_bytes_upload",
                        formats: [
                            {
                                format: "number",
                                factor: 0.000001
                            },
                            {
                                format: "speed"
                            }
                        ],
                        unit: "MB"
                    },
                    {
                        key: "testdl_if_bytes_download",
                        formats: [
                            {
                                format: "number",
                                factor: 0.000001
                            },
                            {
                                format: "speed"
                            }
                        ],
                        unit: "MB"
                    },
                    {
                        key: "testdl_if_bytes_upload",
                        formats: [
                            {
                                format: "number",
                                factor: 0.000001
                            },
                            {
                                format: "speed"
                            }
                        ],
                        unit: "MB"
                    },
                    {
                        key: "testul_if_bytes_download",
                        formats: [
                            {
                                format: "number",
                                factor: 0.000001
                            },
                            {
                                format: "speed"
                            }
                        ],
                        unit: "MB"
                    },
                    {
                        key: "testul_if_bytes_upload",
                        formats: [
                            {
                                format: "number",
                                factor: 0.000001
                            },
                            {
                                format: "speed"
                            }
                        ],
                        unit: "MB"
                    },
                    {
                        key: "time",
                        formats: [
                            {
                                format: "utc2local"
                            }
                        ]
                    },
                    {
                        key: "time_dl_ms",
                        formats: [
                            {
                                format: "ping"
                            }
                        ],
                        unit: "ms"
                    },
                    {
                        key: "time_ul_ms",
                        formats: [
                            {
                                format: "ping"
                            }
                        ],
                        unit: "ms"
                    },
                    {
                        key: "upload_classification"
                    },
                    {
                        key: "upload_kbit",
                        formats: [
                            {
                                format: "number",
                                factor: 0.001
                            },
                            {
                                format: "speed"
                            }
                        ],
                        unit: "Mbit/s"
                    },
                    {
                        key: "wifi_link_speed",
                        unit: "Mbit/s"
                    },
                    {
                        key: "zip_code"
                    }
                ]
            }
        ],
        view: {
            zoom_initial: 15,
            zoom_min: 4,
            zoom_max: 18,
            default_map: "google",
            map_provider: ["google", "osm"]
        },
        graph: {
            step_ms: 10
        }
    },
    user: {
        allow_query_uuid: true,
        allow_set_uuid: false,
        shown: {
            force_ip4: false,
            invisible: false,
            anonymous_mode: true,
            no_anonymize_before_delete_user: false,
            delete_user: true,
            client_uuid: true
        }
    }
};
